### PR TITLE
memory / logging and order feed fixes

### PR DESCRIPTION
**Pull request by pcGarethJames**
- improvements to memory usage in feeds and deltas
 - changed feed error logging so that it goes to a proper log location
 - fixed a bug with the feed status panel which prevented it from showing an error status
 - added a new debug logging option to feeds and deltas, do give more detailed information about the process when it runs
 - Fixes to order feed to prevent errors due to unexpected data formats
 - Tweaked the data ranges for user & order feeds so they only pull in the most relevant data

### CODE REVIEW

File: includes/admin/class-pureclarity-settings-page.php [hunk 1]
```diff
@@ -147,6 +147,15 @@ private function get_general_fields() {
 			'value_cb'    => 'is_nightly_feed_enabled',
 		);
 
+		$feeds_logging_enabled_checkbox = array(
+			'name'        => 'pureclarity_feed_debug_logging',
+			'label_for'   => 'pureclarity_feed_debug_logging',
+			'label'       => 'Enable Debug Logging',
+			'type'        => 'checkbox',
+			'description' => 'Check to enable extra logging for feeds & deltas, to show what items are being process / skipped. Please only enable as a temporary measure, as it can produce large log files.',
```
Comment (jcbarton) [inline includes/admin/class-pureclarity-settings-page.php]: process**ed**

**Review by PeterBrooksbank — APPROVED** [review]
LGTM

File: includes/admin/class-pureclarity-settings-page.php (hunk 1 above)
Comment (jcbarton) [reply includes/admin/class-pureclarity-settings-page.php]: please only enable when requested by PureClarity support ?

File: includes/class-pureclarity-order.php [hunk 2]
```diff
@@ -40,7 +40,7 @@ public function get_order_info( $order_id ) {
 						if ( is_object( $product ) ) {
 							$order_info['items'][] = array(
 								'id'        => $item->get_product_id(),
-								'qty'       => $item['qty'],
+								'qty'       => ceil( $item['qty'] ),
```
Comment (jcbarton) [inline includes/class-pureclarity-order.php:43]: just wondering what happens if we dont have qty?

before would have set property to undefined... now will throw error?

File: includes/class-pureclarity-session.php [hunk 3]
```diff
@@ -240,7 +240,7 @@ public function set_cart() {
 			foreach ( $cart_items as $cart_item_key => $cart_item ) {
 				$item    = array(
 					'id'        => $cart_item['product_id'],
-					'qty'       => $cart_item['quantity'],
+					'qty'       => ceil( $cart_item['quantity'] ),
```
Comment (jcbarton) [inline includes/class-pureclarity-session.php:243]: same applies (if it applies!)

File: includes/cron/class-pureclarity-cron-deltas.php [hunk 4]
```diff
@@ -57,11 +57,13 @@ public function __construct(
 	 */
 	public function run_delta_schedule() {
 		if ( false === $this->deltas->is_delta_running() ) {
+			wp_suspend_cache_addition( true );
```
Comment (jcbarton) [inline includes/cron/class-pureclarity-cron-deltas.php]: is this per process? wondering if we need to try/catch + always set back?

File: includes/cron/class-pureclarity-cron-deltas.php [hunk 5]
```diff
@@ -157,9 +170,38 @@ public function process_users() {
...
+	/**
+	 * Logs an error using WooCommerce Logging.
+	 *
+	 * @param string $type - delta type.
+	 * @param string $message - error message.
+	 */
+	private function log_error( $type, $message ) {
+		$logger = wc_get_logger();
+		if ( $logger ) {
+			$logger->error( "PureClarity {$type} delta error: {$message}" );
```
Comment (jcbarton) [inline includes/cron/class-pureclarity-cron-deltas.php:195]: delta error?

File: includes/cron/class-pureclarity-cron-deltas.php [hunk 6]
```diff
@@ -157,9 +170,38 @@ public function process_users() {
...
+	 * Logs an error using WooCommerce Logging.
+	 *
+	 * @param string $type - delta type.
+	 * @param string $message - error message.
+	 */
+	private function log_debug( $type, $message ) {
+		if ( $this->settings->is_feed_logging_enabled() ) {
+			$logger = wc_get_logger();
+			if ( $logger ) {
+				$logger->debug( "PureClarity {$type} delta debug: {$message}" );
```
Comment (jcbarton) [inline includes/cron/class-pureclarity-cron-deltas.php:209]: same here - is this a general function?

File: includes/feeds/class-pureclarity-feed.php [hunk 7]
```diff
@@ -576,8 +591,19 @@ public function get_categories() {
 	 */
 	public function get_users_count() {
 		$args = array(
-			'order'   => 'ASC',
-			'orderby' => 'ID',
+			'order'                  => 'ASC',
+			'orderby'                => 'ID',
+			'fields'                 => 'ids',
+			'update_post_meta_cache' => false,
+			'meta_query'             => array( // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
```
Comment (jcbarton) [inline includes/feeds/class-pureclarity-feed.php:606]: commented out?

File: includes/feeds/class-pureclarity-feed.php [hunk 8]
```diff
@@ -595,83 +621,95 @@ public function get_users_count() {
...
-			'order'   => 'ASC',
-			'orderby' => 'ID',
-			'offset'  => $page_size * ( $current_page - 1 ),
-			'number'  => $page_size,
+			'order'                  => 'ASC',
+			'orderby'                => 'ID',
+			'offset'                 => $page_size * ( $current_page - 1 ),
+			'number'                 => $page_size,
+			'update_post_meta_cache' => false,
+			'meta_query'             => array( // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
```
Comment (jcbarton) [inline includes/feeds/class-pureclarity-feed.php:640]: ..and here

File: includes/admin/class-pureclarity-settings-page.php (hunk 1 above)
Comment (pcGarethJames) [reply includes/admin/class-pureclarity-settings-page.php]: how about "Please only enable if you are troubleshooting the feeds/deltas", and I'll ass some info tot he troubleshooting docs to recommend using it for certain things? (means we can promote independence :) )

**Pull request merged**